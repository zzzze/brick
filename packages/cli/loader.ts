import { pattern } from '@brick/shared'
const flagMap: Record<string, string> = {}

function setFuncFlag(value: unknown, label = 'brickroot', path = ''): unknown {
  if (!value) {
    return value
  }
  if (label === 'children' && Array.isArray(value)) {
    return value.map(function (item) {
      return setFuncFlag(item)
    })
  }
  if (typeof value === 'object') {
    return Object.keys(value || {}).reduce<Record<string, unknown>>(function (result, key) {
      result[key] = setFuncFlag(result[key], key, path ? `${path}-${label}` : label)
      return result
    }, (value as Record<string, unknown>) || {})
  }
  if (path === 'brickroot' && label === 'render' && typeof value === 'string') {
    const key = Math.random().toString(36).slice(2).toUpperCase()
    flagMap[key] = `function (components) {
      return ${value}
    }`
    return key
  }
  if (
    ['brickroot-supply-actions', 'brickroot-actions', 'brickroot-handlers', 'brickroot-listeners'].includes(path) &&
    typeof value === 'string' &&
    pattern.FUNCTION_STR.test(value)
  ) {
    const key = Math.random().toString(36).slice(2).toUpperCase()
    flagMap[key] = `function (setData, emit) {
      return ${value}
    }`
    return key
  }
  return value
}

export default function loader(source: string | Record<string, unknown>): string {
  // eslint-disable-next-line
  // @ts-ignore
  if (this.cacheable) this.cacheable() // eslint-disable-line
  let obj: Record<string, unknown> =
    typeof source === 'string' ? (JSON.parse(source) as Record<string, unknown>) : source
  obj = setFuncFlag(obj) as Record<string, unknown>
  let value = JSON.stringify(obj, null, 2)
    .replace(/\u2028/g, '\\u2028')
    .replace(/\u2029/g, '\\u2029')
  Object.keys(flagMap).forEach(function (key) {
    value = value.replace(`"${key}"`, flagMap[key])
  })
  return `module.exports = ${value}`
}
