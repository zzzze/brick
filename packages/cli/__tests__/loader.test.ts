import loader from '../loader'

const config = `
{
  "name": "View",
  "id": "container",
  "data": {
    "name": "bazx"
  },
  "supply": {
    "data": {
      "text": "{{$this.name}}"
    },
    "actions": {
      "onClick": "name => {\\n            setData(function(data) {\\n    console.log('000001', data);\\n          return Object.assign({}, data, {\\n                name: name,\\n              })\\n            }, {\\n              setToBlueprint: true,\\n            })\\n          }"
    }
  },
  "handlers": {
    "handler-01": "() => console.log('hello')"
  },
  "listeners": {
    "listener-01": "() => console.log('world')"
  },
  "render": "args => <div>abc</div>",
  "children": [
    {
      "name": "Text",
      "data": {
        "content": "{{$supply.$container.text}}"
      },
      "version": "0.0.1"
    },
    {
      "name": "Text",
      "data": {
        "content": "fooo"
      },
      "version": "0.0.1"
    },
    {
      "name": "View",
      "id": "container2",
      "supply": {
        "actions": {
          "onClick": "{{$supply.$container.onClick}}"
        }
      },
      "children": [
        {
          "name": "TextWithAction",
          "actions": {
            "click": "function($this, $supply) {\\n window.xx = $supply;\\n                 $supply.actions.$container2.onClick($this.data.content)\\n                }",
            "handleClick": "function($this, $supply) {\\n                  $this.actions.click($this, $supply)\\n                }"
          },
          "handlers": {
            "onClick": "{{$this.handleClick}}"
          },
          "data": {
            "content": "123456789"
          },
          "version": "0.0.1"
        }
      ],
      "version": "0.0.1"
    },
    {
      "name": "View",
      "children": [
        {
          "name": "Text",
          "data": {
            "content": "{{$supply.$container.text}}"
          },
          "version": "0.0.1"
        }
      ],
      "version": "0.0.1"
    }
  ],
  "version": "0.0.1"
}
`

test('load module correctly', () => {
  const fakeWebpack = {
    cacheable: false,
  }
  expect(loader.call(fakeWebpack, config)).toMatchSnapshot()
})
