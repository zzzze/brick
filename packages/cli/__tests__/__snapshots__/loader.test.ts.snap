// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`load module correctly 1`] = `
"module.exports = {
  \\"name\\": \\"View\\",
  \\"id\\": \\"container\\",
  \\"data\\": {
    \\"name\\": \\"bazx\\"
  },
  \\"supply\\": {
    \\"data\\": {
      \\"text\\": \\"{{$this.name}}\\"
    },
    \\"actions\\": {
      \\"onClick\\": function (setData, emit) {
      return name => {
            setData(function(data) {
    console.log('000001', data);
          return Object.assign({}, data, {
                name: name,
              })
            }, {
              setToBlueprint: true,
            })
          }
    }
    }
  },
  \\"handlers\\": {
    \\"handler-01\\": function (setData, emit) {
      return () => console.log('hello')
    }
  },
  \\"listeners\\": {
    \\"listener-01\\": function (setData, emit) {
      return () => console.log('world')
    }
  },
  \\"render\\": function (components) {
      return args => <div>abc</div>
    },
  \\"children\\": [
    {
      \\"name\\": \\"Text\\",
      \\"data\\": {
        \\"content\\": \\"{{$supply.$container.text}}\\"
      },
      \\"version\\": \\"0.0.1\\"
    },
    {
      \\"name\\": \\"Text\\",
      \\"data\\": {
        \\"content\\": \\"fooo\\"
      },
      \\"version\\": \\"0.0.1\\"
    },
    {
      \\"name\\": \\"View\\",
      \\"id\\": \\"container2\\",
      \\"supply\\": {
        \\"actions\\": {
          \\"onClick\\": \\"{{$supply.$container.onClick}}\\"
        }
      },
      \\"children\\": [
        {
          \\"name\\": \\"TextWithAction\\",
          \\"actions\\": {
            \\"click\\": function (setData, emit) {
      return function($this, $supply) {
 window.xx = $supply;
                 $supply.actions.$container2.onClick($this.data.content)
                }
    },
            \\"handleClick\\": function (setData, emit) {
      return function($this, $supply) {
                  $this.actions.click($this, $supply)
                }
    }
          },
          \\"handlers\\": {
            \\"onClick\\": \\"{{$this.handleClick}}\\"
          },
          \\"data\\": {
            \\"content\\": \\"123456789\\"
          },
          \\"version\\": \\"0.0.1\\"
        }
      ],
      \\"version\\": \\"0.0.1\\"
    },
    {
      \\"name\\": \\"View\\",
      \\"children\\": [
        {
          \\"name\\": \\"Text\\",
          \\"data\\": {
            \\"content\\": \\"{{$supply.$container.text}}\\"
          },
          \\"version\\": \\"0.0.1\\"
        }
      ],
      \\"version\\": \\"0.0.1\\"
    }
  ],
  \\"version\\": \\"0.0.1\\"
}"
`;
